### PR TITLE
Split cmd.py into smaller modules

### LINKED ISSUE #27: Get rid of global state in the implementation
As for now it should be enough to extract the following classes:
- `GitContext`, as a lazily evaluated cache of config, logs, reflogs, merge-bases etc.
- `MacheteContext`, for storing the branch dependency tree structure
- `CommandLineContext`, for options/command/args
- possibly `SystemContext` for checking/caching things like `isatty`, hook executability etc.

### LINKED ISSUE #34: Split the main file (cmd.py) into smaller modules
~It remains an open question if/how to support the old installation via Makefile.~ => we have now better ways of installing git-machete, no longer a problem

### PR DESCRIPTION
**Pull request by rifel123**
implements issue #34

### CODE REVIEW

File: git_machete/cmd.py [hunk 1]
```diff
@@ -1636,7 +1371,7 @@ def log_result() -> Generator[str, None, None]:
                     def branch_def_to_str(lb: str, lb_or_rb: str) -> str:
                         return lb if lb == lb_or_rb else f"{lb_or_rb} (remote counterpart of {lb})"
 
-                    joined_branch_defs = ", ".join(map(tupled(branch_def_to_str), branch_defs))
+                    joined_branch_defs = ", ".join(map(utils.tupled(branch_def_to_str), branch_defs))
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Nit: `tupled`, `get_second` and other low-level utils can probably be `from`-imported

File: git_machete/cmd.py [hunk 2]
```diff
@@ -10,147 +10,23 @@
...
 import sys
 import textwrap
 
+from git_machete import utils
+from git_machete.utils import RED, YELLOW, ORANGE, GREEN, DIM, BOLD, ENDC, dim, pretty_choices, bold, colored, debug, fmt, underline
+from git_machete.contexts import CommandLineContext
+from git_machete.exceptions import MacheteException, StopTraversal
+from git_machete.docs import short_docs, long_docs
+from git_machete.git_operations import GitContext, REFLOG_ENTRY
+from git_machete import git_operations
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Given how often this module is referenced... maybe just `git`? or will it clash with sth else this way?

File: git_machete/cmd.py (hunk 1 above)
Comment (PawelLipski) [reply git_machete/cmd.py]: Same for e.g. `flat_map`, `excluding`

File: git_machete/cmd.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
```
Comment (PawelLipski) [inline git_machete/cmd.py:2]: I'd rename this entire file to sth like `main.py`... can be on a separate PR tho (since changes in other places where `cmd` is referenced will probably be needed as well)

File: git_machete/cmd.py [hunk 4]
```diff
@@ -10,147 +10,23 @@
 import os
 import re
 import shutil
-import subprocess
 import sys
 import textwrap
 
+from git_machete import utils
+from git_machete.utils import RED, YELLOW, ORANGE, GREEN, DIM, BOLD, ENDC, dim, pretty_choices, bold, colored, debug, fmt, underline
+from git_machete.contexts import CommandLineContext
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Hmm I don't like the `contexts` subpackage name, it's extremely non-specific... lemme think of better alternatives...

File: git_machete/cmd.py [hunk 5]
```diff
@@ -196,145 +57,19 @@ def pick(choices: List[str], name: str, apply_fmt: bool = True) -> str:
...
-def debug(cli_ctxt: CommandLineContext, hdr: str, msg: str) -> None:
-    if cli_ctxt.opt_debug:
-        sys.stderr.write(f"{bold(hdr)}: {dim(msg)}\n")
-
-
 # To avoid displaying the same warning multiple times during a single run.
 displayed_warnings: Set[str] = set()
 
 
 def warn(msg: str, apply_fmt: bool = True) -> None:
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Hmmm intuitively I'd say that `debug` and `warn` should be in the very same place (probably `utils`, not this file)

Comment (PawelLipski) [reply git_machete/cmd.py]: Coz they are conceptually just different levels of logging

File: git_machete/cmd.py [hunk 6]
```diff
@@ -196,145 +57,19 @@ def pick(choices: List[str], name: str, apply_fmt: bool = True) -> str:
...
-
-def current_directory_or_none() -> Optional[str]:
-    try:
-        return os.getcwd()
-    except OSError:
-        # This happens when current directory does not exist (typically: has been deleted)
-        return None
-
-
 # Let's keep the flag to avoid checking for current directory's existence
```
Comment (PawelLipski) [inline git_machete/cmd.py]: This comment applies to `current_directory_confirmed_to_exist`, should be moved together

File: git_machete/cmd.py [hunk 7]
```diff
@@ -2263,95 +1662,34 @@ def contains_equivalent_tree(
...
-
 
 # Since getting the full history of a branch can be an expensive operation for large repositories (compared to all other underlying git operations),
 # there's a simple optimization in place: we first fetch only a couple of first commits in the history,
 # and only fetch the rest if none of them occurs on reflog of any other branch.
 def spoonfeed_log_shas(cli_ctxt: CommandLineContext, b: str) -> Generator[str, None, None]:
-    if b not in initial_log_shas_cached:
-        initial_log_shas_cached[b] = log_shas(cli_ctxt, b, max_count=MAX_COUNT_FOR_INITIAL_LOG)
-    for sha in initial_log_shas_cached[b]:
+    if b not in GitContext.initial_log_shas_cached:
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Hmm I think this method (`spoonfeed_log_shas`) is Machete-agnostic enough that it can go directly to `GitContext`

File: git_machete/cmd.py [hunk 8]
```diff
@@ -2486,47 +1820,41 @@ def get_relation_to_remote_counterpart(cli_ctxt: CommandLineContext, b: str, rb:
...
     # %gs - reflog subject
-    all_branches = [f"refs/heads/{b}" for b in local_branches(cli_ctxt)] + \
-                   [f"refs/remotes/{combined_counterpart_for_fetching_of_branch(cli_ctxt, b)}" for b in local_branches(cli_ctxt) if combined_counterpart_for_fetching_of_branch(cli_ctxt, b)]
+    all_branches = [f"refs/heads/{b}" for b in git_operations.local_branches(cli_ctxt)] + \
+                   [f"refs/remotes/{git_operations.combined_counterpart_for_fetching_of_branch(cli_ctxt, b)}" for b in git_operations.local_branches(cli_ctxt) if git_operations.combined_counterpart_for_fetching_of_branch(cli_ctxt, b)]
     # The trailing '--' is necessary to avoid ambiguity in case there is a file called just exactly like one of the branches.
-    entries = non_empty_lines(popen_git(cli_ctxt, "reflog", "show", "--format=%gD\t%H\t%gs", *(all_branches + ["--"])))
-    reflogs_cached = {}
+    entries = utils.non_empty_lines(git_operations.popen_git(cli_ctxt, "reflog", "show", "--format=%gD\t%H\t%gs", *(all_branches + ["--"])))
+    GitContext.reflogs_cached = {}
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Hmmm that's a code smell &mdash; things like a cache should be an implementation detail of `GitContext`, **not** exposed to the outside of `GitContext`.

Other than making fields like `reflogs_cached` private (which can be done in one go on a separate PR), I'd consider moving this entire `load_all_reflogs` method directly to `GitContext` &mdash; there's nothing inherently Machete-specific here AFAICS.

File: git_machete/cmd.py [hunk 9]
```diff
@@ -2536,31 +1864,30 @@ def load_all_reflogs(cli_ctxt: CommandLineContext) -> None:
...
         b, pos = branch_and_pos
-        if b not in reflogs_cached:
-            reflogs_cached[b] = []
-        reflogs_cached[b] += [(sha, subject)]
+        if b not in GitContext.reflogs_cached:
+            GitContext.reflogs_cached[b] = []
+        GitContext.reflogs_cached[b] += [(sha, subject)]
 
 
 def reflog(cli_ctxt: CommandLineContext, b: str) -> List[REFLOG_ENTRY]:
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Same here, this method probably belongs to `GitContext` as well, reflogs are by no means machete-specific

File: git_machete/cmd.py [hunk 10]
```diff
@@ -2536,31 +1864,30 @@ def load_all_reflogs(cli_ctxt: CommandLineContext) -> None:
...
+                tuple(entry.split(":", 1)) for entry in utils.non_empty_lines(  # type: ignore
                     # The trailing '--' is necessary to avoid ambiguity in case there is a file called just exactly like the branch 'b'.
-                    popen_git(cli_ctxt, "reflog", "show", "--format=%H:%gs", b, "--"))
+                    git_operations.popen_git(cli_ctxt, "reflog", "show", "--format=%H:%gs", b, "--"))
             ]
-        return reflogs_cached[b]
+        return GitContext.reflogs_cached[b]
 
 
 def get_latest_checkout_timestamps(cli_ctxt: CommandLineContext) -> Dict[str, int]:  # TODO (#110): default dict with 0
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Ditto, pretty unusual, but not machete-specific either

File: git_machete/cmd.py [hunk 11]
```diff
@@ -2606,34 +1933,27 @@ def is_merged_to(machete_client: MacheteClient, cli_ctxt: CommandLineContext, b:
 
 
 def get_revision_repr(cli_ctxt: CommandLineContext, revision: str) -> str:
```
Comment (PawelLipski) [inline git_machete/cmd.py]: Ditto

Comment (PawelLipski) [reply git_machete/cmd.py]: Btw, I've got an impression that in 90% of methods that accept `CommandLineContext`, it's only ever used to access `opt_debug`... actually I'd consider making `opt_debug` a global variable again to simplify things 😅 to be done on a separate PR

File: git_machete/cmd.py [hunk 12]
```diff
@@ -3696,7 +2365,7 @@ def check_required_param(in_args: List[str], allowed_values_string: str) -> str:
             cb = current_branch(cli_ctxt)
             dest = machete_client.parse_direction(param, cb, allow_current=False, down_pick_mode=True)
             if dest != cb:
-                go(cli_ctxt, dest)
+                git_operations.go(cli_ctxt, dest)
```
Comment (PawelLipski) [inline git_machete/cmd.py]: While you're here, I'd change the name of this method from `go` (which is a machete-specific term) to `checkout` (which is machete-agnostic)

File: git_machete/contexts.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+from typing import Optional, List
+
+
+class CommandLineContext:
```
Comment (PawelLipski) [inline git_machete/contexts.py]: Hmmm this class name doesn't clearly indicate what this class does... although that's probably gonna be massively reworked soon anyway within the scope of https://github.com/VirtusLab/git-machete/issues/158 and https://github.com/VirtusLab/git-machete/issues/161

Comment (PawelLipski) [reply git_machete/contexts.py]: Anywa, I'd just rename it to `CommandLineOptions` as for now, further steps TBD

File: git_machete/contexts.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        self.opt_roots: List[str] = list()
+        self.opt_start_from: str = "here"
+        self.opt_stat: bool = False
+        self.opt_sync_github_prs: bool = False
+        self.opt_unset_override: bool = False
+        self.opt_verbose: bool = False
+        self.opt_yes: bool = False
+
+
+class SystemContext:
```
Comment (PawelLipski) [inline git_machete/contexts.py]: Hmm is a WIP? or can it be removed?

File: git_machete/docs.py [hunk 15]
```diff
@@ -0,0 +1,654 @@
+from typing import Dict
+
+DISCOVER_DEFAULT_FRESH_BRANCH_COUNT = 10  # make variable declaration in one file only
```
Comment (PawelLipski) [inline git_machete/docs.py]: That's a TODO for this PR right?

File: git_machete/git_operations.py [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+    remote_branches_cached: Optional[List[str]] = None
+    initial_log_shas_cached: Dict[str, List[str]] = {}
+    remaining_log_shas_cached: Dict[str, List[str]] = {}
+    reflogs_cached: Optional[Dict[str, Optional[List[REFLOG_ENTRY]]]] = None
+
+    def __init__(self) -> None:
+        pass
+
+
+def run_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs: Dict[str, str]) -> int:
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: Hmmm would it make sense to, instead of keeping all these function global, convert them into instance methods of `GitContext`?

Comment (PawelLipski) [reply git_machete/git_operations.py]: This way, instead of keeping things like `config_cached` as class attributes (i.e. de facto still **global** variables), you could move them into instance attributes of `GitContext`

File: git_machete/git_operations.py [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+from git_machete.contexts import CommandLineContext
+from git_machete.exceptions import MacheteException
+from git_machete import utils
+
+REFLOG_ENTRY = Tuple[str, str]
+
+
+class GitContext:
+
+    counterparts_for_fetching_cached: Optional[Dict[str, Optional[str]]] = None  # TODO (#110): default dict with None
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: Move next to `fetch_done_for` for logical grouping

File: git_machete/utils.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
+
+from typing import Any, Callable, Iterable, List, Optional, Tuple, TypeVar, Dict
+
+import os
+import sys
```
Comment (PawelLipski) [inline git_machete/utils.py:5]: Nit: keep imports alphabetical

File: git_machete/utils.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+        elif c.lower() == 'yq':
+            return colored(c[0], GREEN) + colored(c[1], RED)
+        elif c.lower() in ('n', 'q'):
+            return colored(c, RED)
+        else:
+            return colored(c, ORANGE)
+    return f" ({', '.join(map_truthy_only(format_choice, choices))}) "
+
+
+def debug(cli_ctxt: CommandLineContext, hdr: str, msg: str) -> None:
```
Comment (PawelLipski) [inline git_machete/utils.py]: Yeah, since the only attributes of `CommandLineContext` accessed in `utils` are `opt_debug` and `opt_verbose`, I'd consider making them global variables for the sake of simplicity (hence `cli_ctxt` param will no longer be needed in many methods across other files as well)

File: git_machete/cmd.py (hunk 2 above)
Comment (PawelLipski) [reply git_machete/cmd.py]: Okay, probably this comment will no longer be relevant once `GitContext` is extracted (since you'll rather use `GitContext` instance rather than reference the module directly)

File: git_machete/constants.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+import os
+
+MAX_COUNT_FOR_INITIAL_LOG = 10
+DISCOVER_DEFAULT_FRESH_BRANCH_COUNT = 10
+
+PICK_FIRST_ROOT: int = 0
+PICK_LAST_ROOT: int = -1
+
+BRANCH_DEF = Tuple[str, str]
```
Comment (PawelLipski) [inline git_machete/constants.py]: Hmm type alias doesn't feel right here among there numeric/string constants, I'd rather put it back to cmd.py

File: git_machete/git_operations.py [hunk 21]
```diff
@@ -376,3 +375,24 @@ def load_branches(cli_ctxt: CommandLineContext) -> None:
...
+
+
+def log_shas(cli_ctxt: CommandLineContext, revision: str, max_count: Optional[int]) -> List[str]:
+    opts = ([f"--max-count={str(max_count)}"] if max_count else []) + ["--format=%H", f"refs/heads/{revision}"]
+    return utils.non_empty_lines(popen_git(cli_ctxt, "log", *opts))
+
+
+# Since getting the full history of a branch can be an expensive operation for large repositories (compared to all other underlying git operations),
+# there's a simple optimization in place: we first fetch only a couple of first commits in the history,
+# and only fetch the rest if none of them occurs on reflog of any other branch.
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: I'd rephrase this particular line to sth like:
```suggestion
# and only fetch the rest if needed.
```
to avoid machete-specific mention about reflogs.

Comment (PawelLipski) [reply git_machete/git_operations.py]: https://virtuslab.github.io/tips/#git/git-reflog ➕ https://github.com/VirtusLab/git-machete/blob/master/docs/blog-1/blog-1.md#appendix-fork-point---not-so-easy-business for details (although I admit the latter is pretty technical)

File: git_machete/utils.py [hunk 22]
```diff
@@ -190,13 +190,13 @@ def chdir_upwards_until_current_directory_exists(cli_ctxt: CommandLineContext) -
...
                 current_directory = current_directory_or_none()
-            debug(cli_ctxt,
+            debug(cli_opts,
                   "chdir_upwards_until_current_directory_exists()",
                   f"current directory did not exist, chdired up into {current_directory}")
         current_directory_confirmed_to_exist = True
 
 
-def popen_cmd(cli_ctxt: CommandLineContext, cmd: str, *args: str, **kwargs: Any) -> Tuple[int, str, str]:
+def popen_cmd(cli_ctxt: CommandLineOptions, cmd: str, *args: str, **kwargs: Any) -> Tuple[int, str, str]:
```
Comment (PawelLipski) [inline git_machete/utils.py]: `s/cli_ctxt/cli_opts`, to use [`sed`](https://www.gnu.org/software/sed/manual/html_node/The-_0022s_0022-Command.html) notation ;)

File: git_machete/tests/functional/test_machete.py [hunk 23]
```diff
@@ -79,7 +84,7 @@ def launch_command(*args: str) -> str:
         with io.StringIO() as out:
             with redirect_stdout(out):
                 cmd.launch(list(args))
-                cmd.flush_caches()
+                git_ctxt.flush_caches()
```
Comment (PawelLipski) [inline git_machete/tests/functional/test_machete.py]: Maybe just `git` instead of `git_ctxt`? same in cmd.py

This way it'll sound roughly as in command line: e.g. `git chechout <branch>` vs `git.checkout("<branch>")`

File: git_machete/utils.py (hunk 22 above)
Comment (PawelLipski) [reply git_machete/utils.py]: (here and in a few other places where it's still left)

File: git_machete/git_operations.py [hunk 24]
```diff
@@ -0,0 +1,412 @@
...
+                # %H - full hash
+                # %gs - reflog subject
+                self.reflogs_cached[b] = [
+                    tuple(entry.split(":", 1)) for entry in utils.non_empty_lines(  # type: ignore
+                        # The trailing '--' is necessary to avoid ambiguity in case there is a file called just exactly like the branch 'b'.
+                        self.popen_git("reflog", "show", "--format=%H:%gs", b, "--"))
+                ]
+            return self.reflogs_cached[b]
+
+    def create_branch(self, b: str, out_of_revision: str) -> None:  # maybe move such a methods into github.py
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: Hmmm why github.py? 🤔 this method is independent from GitHub, it still makes sense wherever the remote repository is hosted... or even when there is no remote repo configured at all

Comment (PawelLipski) [reply git_machete/git_operations.py]: The only methods that need to be kept under github.py are the ones that strictly talk to GitHub API, handling things that are **not** known to `git` as such &mdash; like pull requests.

File: git_machete/git_operations.py [hunk 25]
```diff
@@ -0,0 +1,412 @@
...
+            self.initial_log_shas_cached[b] = self.log_shas(b, max_count=MAX_COUNT_FOR_INITIAL_LOG)
+        for sha in self.initial_log_shas_cached[b]:
+            yield sha
+
+        if b not in self.remaining_log_shas_cached:
+            self.remaining_log_shas_cached[b] = self.log_shas(b, max_count=None)[MAX_COUNT_FOR_INITIAL_LOG:]
+        for sha in self.remaining_log_shas_cached[b]:
+            yield sha
+
+    # Reflog magic
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: This line can probably be removed :)

File: git_machete/git_operations.py [hunk 26]
```diff
@@ -0,0 +1,412 @@
...
+from git_machete import utils
+from git_machete.constants import MAX_COUNT_FOR_INITIAL_LOG
+
+REFLOG_ENTRY = Tuple[str, str]
+
+
+class GitContext:
+
+    def __init__(self, cli_opts: CommandLineOptions) -> None:
+        self.cli_opts = cli_opts
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: Add type hint to this attr

File: git_machete/git_operations.py (hunk 24 above)
Comment (rifel123) [reply git_machete/git_operations.py]: Hmm, as every method in git_operations belongs now to GitContext, this comment is outdated :)

File: git_machete/cmd.py [hunk 27]
```diff
@@ -569,36 +282,36 @@ def print_annotation(self, b: str) -> None:
             print(self.annotations[b])
 
     def update(self) -> None:
-        cb = current_branch(self.cli_ctxt)
-        if self.cli_ctxt.opt_merge:
+        cb = current_branch(self.git)
```
Comment (PawelLipski) [inline git_machete/cmd.py:285]: Hmm for a follow-up PR: `current_branch` can probably be a method of `git`

Comment (PawelLipski) [reply git_machete/cmd.py:285]: Hmm actually I can see quite a few methods in cmd.py that **rely on nothing machete-specific**: at first glance, I'd say everything b/w `currently_rebased_branch_or_none` and `get_latest_checkout_timestamps`, inclusive

File: git_machete/cmd.py [hunk 28]
```diff
@@ -902,143 +605,134 @@ def print_new_line(new_status: bool) -> None:
...
-                        f"Branch {bold(b)} diverged from (and has older commits than) its remote counterpart {bold(rb)}.\n"
-                        f"Reset branch {bold(b)} to the commit pointed by {bold(rb)}?" + pretty_choices('y', 'N', 'q',
-                                                                                                        'yq'),
-                        f"Branch {bold(b)} diverged from (and has older commits than) its remote counterpart {bold(rb)}.\n"
-                        f"Resetting branch {bold(b)} to the commit pointed by {bold(rb)}..."
-                    )
+                    rb = self.git.strict_counterpart_for_fetching_of_branch(b)
+                    ans = ask_if(self.cli_opts,
+                                 f"Branch {bold(b)} diverged from (and has older commits than) its remote counterpart {bold(rb)}.\n"
+                                 f"Reset branch {bold(b)} to the commit pointed by {bold(rb)}?" + pretty_choices('y',
```
Comment (PawelLipski) [inline git_machete/cmd.py:684]: Whoops awkward formatting... maybe just merge into a single line

File: git_machete/cmd.py [hunk 29]
```diff
@@ -902,143 +605,134 @@ def print_new_line(new_status: bool) -> None:
...
-                        f"Branch {bold(b)} diverged from (and has newer commits than) its remote counterpart {bold(rb)}.\n"
-                        f"Push {bold(b)} with force-with-lease to {bold(remote)}?" + pretty_choices('y', 'N', 'q', 'yq'),
-                        f"Branch {bold(b)} diverged from (and has newer commits than) its remote counterpart {bold(rb)}.\n"
-                        f"Pushing {bold(b)} with force-with-lease to {bold(remote)}...",
-                        override_answer=None if self.cli_ctxt.opt_push_tracked else "N"
-                    )
+                    rb = self.git.strict_counterpart_for_fetching_of_branch(b)
+                    ans = ask_if(self.cli_opts,
+                                 f"Branch {bold(b)} diverged from (and has newer commits than) its remote counterpart {bold(rb)}.\n"
+                                 f"Push {bold(b)} with force-with-lease to {bold(remote)}?" + pretty_choices('y', 'N',
```
Comment (PawelLipski) [inline git_machete/cmd.py:703]: Nit: merge next line

File: git_machete/cmd.py [hunk 30]
```diff
@@ -2587,53 +1774,34 @@ def get_latest_checkout_timestamps(cli_ctxt: CommandLineContext) -> Dict[str, in
 
 # Complex routines/commands
 
-def is_merged_to(machete_client: MacheteClient, cli_ctxt: CommandLineContext, b: str, target: str) -> bool:
-    if is_ancestor_or_equal(cli_ctxt, b, target):
+def is_merged_to(machete_client: MacheteClient, b: str, target: str) -> bool:
```
Comment (PawelLipski) [inline git_machete/cmd.py:1777]: Hmmm shouldn't this one live in `MacheteClient`?

File: git_machete/cmd.py [hunk 31]
```diff
@@ -2587,53 +1774,34 @@ def get_latest_checkout_timestamps(cli_ctxt: CommandLineContext) -> Dict[str, in
...
         # If there is a commit in target with an identical tree state to b,
         # then b may be squash or rebase merged into target.
-        return contains_equivalent_tree(cli_ctxt, b, target)
+        return contains_equivalent_tree(machete_client.git, b, target)
 
 
-def get_revision_repr(cli_ctxt: CommandLineContext, revision: str) -> str:
-    short_sha = short_commit_sha_by_revision(cli_ctxt, revision)
-    if is_full_sha(revision) or revision == short_sha:
+def get_revision_repr(git: GitContext, revision: str) -> str:
```
Comment (PawelLipski) [inline git_machete/cmd.py:1795]: This one probably in `GitContext` (although a bit of a gray area)

File: git_machete/cmd.py [hunk 32]
```diff
@@ -2587,53 +1774,34 @@ def get_latest_checkout_timestamps(cli_ctxt: CommandLineContext) -> Dict[str, in
...
-    initial_log_shas_cached = {}
-    local_branches_cached = None
-    reflogs_cached = None
-    remaining_log_shas_cached = {}
-    remote_branches_cached = None
-
-
-def pick_remote(cli_ctxt: CommandLineContext, b: str) -> None:
-    rems = remotes(cli_ctxt)
+def pick_remote(git: GitContext, b: str) -> None:
```
Comment (PawelLipski) [inline git_machete/cmd.py:1803]: Move into `MacheteClient` (this could be a private, used only by `traverse` under the hood)

File: git_machete/cmd.py [hunk 33]
```diff
@@ -2645,33 +1813,33 @@ def pick_remote(cli_ctxt: CommandLineContext, b: str) -> None:
...
             raise MacheteException(f"Invalid index: {index + 1}")
-        handle_untracked_branch(cli_ctxt, rems[index], b)
+        handle_untracked_branch(git, rems[index], b)
     except ValueError:
         pass
 
 
-def handle_untracked_branch(cli_ctxt: CommandLineContext, new_remote: str, b: str) -> None:
-    rems: List[str] = remotes(cli_ctxt)
+def handle_untracked_branch(git: GitContext, new_remote: str, b: str) -> None:
```
Comment (PawelLipski) [inline git_machete/cmd.py:1821]: Ditto
